hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclHelper.java

```diff
@@ -0,0 +1,745 @@
+        .map(operation -> setSingleHDFSAcl(operation)).collect(Collectors.toList());
+    CompletableFuture<Void> future =
+        CompletableFuture.allOf(futures.toArray(new CompletableFuture[futures.size()]));
+    return future;
+  }
+
+  private static AclEntry aclEntry(AclEntryScope scope, String name, FsAction action) {
+    return new AclEntry.Builder().setScope(scope)
+        .setType(AuthUtil.isGroupPrincipal(name) ? GROUP : USER).setName(name).setPermission(action)
+        .build();
+  }
+
+  /**
+   * Inner class used to describe modify or remove what acl entries for files or directories(and
+   * child files)
+   */
+  private static class HDFSAclOperation {
+    enum OperationType {
+      MODIFY, REMOVE
+    }
+
+    interface Operation {
+      void apply(FileSystem fs, Path path, List<AclEntry> aclList) throws IOException;
+    }
+
+    private FileSystem fs;
+    private Path path;
+    private List<AclEntry> dirAcl;
+    private List<AclEntry> fileAcl;
+    private boolean recursive;
+    private Operation operation;
+    private List<Path> skipPaths;
+
+    HDFSAclOperation(FileSystem fs, Path path, Set<String> users, OperationType operationType,
+        FsAction fsAction, boolean recursive, List<Path> skipPaths) {
+      this.fs = fs;
+      this.path = path;
+      this.dirAcl = getDefaultAclEntries(users, fsAction);
+      this.fileAcl = getAccessAclEntries(users, fsAction);
+      this.recursive = recursive;
+      this.skipPaths = skipPaths;
+      if (operationType == OperationType.MODIFY) {
+        operation =
+            (fileSystem, modifyPath, aclList) -> fileSystem.modifyAclEntries(modifyPath, aclList);
+      } else if (operationType == OperationType.REMOVE) {
+        operation =
+            (fileSystem, modifyPath, aclList) -> fileSystem.removeAclEntries(modifyPath, aclList);
+      }
+    }
+
+    HDFSAclOperation(Path path, HDFSAclOperation parent) {
+      this.fs = parent.fs;
+      this.path = path;
+      this.dirAcl = parent.dirAcl;
+      this.fileAcl = parent.fileAcl;
+      this.operation = parent.operation;
+      this.recursive = parent.recursive;
+      this.skipPaths = parent.skipPaths;
+    }
+
+    List<HDFSAclOperation> getChildAclOperations() throws IOException {
+      List<HDFSAclOperation> hdfsAclOperations = new ArrayList<>();
+      if (!skipPaths.contains(path) && recursive && fs.isDirectory(path)) {
+        FileStatus[] fileStatuses = fs.listStatus(path);
+        for (FileStatus fileStatus : fileStatuses) {
+          hdfsAclOperations.add(new HDFSAclOperation(fileStatus.getPath(), this));
+        }
+      }
+      return hdfsAclOperations;
+    }
+
+    void setAcl() throws IOException {
+      if (!skipPaths.contains(path) && fs.exists(path)) {
+        if (fs.isDirectory(path)) {
+          if (recursive) {
+            operation.apply(fs, path, dirAcl);
+          } else {
+            operation.apply(fs, path, fileAcl);
```

Comment:
When fs.isDirectory is true, then we will apply the fileAcl ?  That's quite strange... I think the dirAcl or fileAcl var name should be refactored  ?

Reply:
renamed to accessAcls and defaultAndAccessAcls